volatility3/framework/symbols/linux/extensions/__init__.py

```diff
@@ -1428,3 +1429,110 @@ def get_type(self):
+    def _get_cred_int_value(self, member: str) -> int:
+        """Helper to obtain the right cred member value for the current kernel.
+
+        Args:
+            member (str): The requested cred member name to obtain its value
+
+        Raises:
+            AttributeError: When the requested cred member doesn't exist
+            AttributeError: When the cred implementation is not supported.
+
+        Returns:
+            int: The cred member value
+        """
+        if not self.has_member(member):
+            raise AttributeError(f"struct cred doesn't have a '{member}' member")
+
+        cred_val = self.member(member)
+        if hasattr(cred_val, "val"):
+            # From kernels 3.5.7 on it is a 'kuid_t' type
+            value = cred_val.val
+        elif isinstance(cred_val, objects.Integer):
+            # From at least 2.6.30 and until 3.5.7 it was a 'uid_t' type which was an 'unsigned int'
+            value = cred_val
+        else:
+            raise AttributeError("Kernel struct cred is not supported")
+
+        return int(value)
+
+    @property
+    def euid(self):
+        """Returns the effective user ID
+
+        Returns:
+            int: the effective user ID value
+        """
+        return self._get_cred_int_value("euid")
+
+
+class kernel_cap_struct(objects.StructType):
+    # struct kernel_cap_struct was added in kernels 2.5.0
+    @classmethod
+    def get_last_cap_value(cls) -> int:
+        """Returns the latest capability ID supported by the framework.
+
+        Returns:
+            int: The latest supported capability ID supported by the framework.
+        """
+        return len(CAPABILITIES) - 1
+
+    @classmethod
+    def capabilities_to_string(cls, capabilities_bitfield: int) -> List[str]:
+        """Translates a capability bitfield to a list of capability strings.
+
+        Args:
+            capabilities_bitfield (int): The capability bitfield value.
+
+        Returns:
+            List[str]: A list of capability strings.
+        """
+
+        capabilities = []
+        for bit, name in enumerate(CAPABILITIES):
+            if capabilities_bitfield & (1 << bit) != 0:
+                capabilities.append(name)
+
+        return capabilities
+
+    def get_capabilities(self) -> int:
+        """Returns the capability bitfield value
+
+        Returns:
+            int: The capability bitfield value.
+        """
+        # In kernels 2.6.25.20 the kernel_cap_struct::cap became and array
+        cap_value = self.cap[0] if isinstance(self.cap, objects.Array) else self.cap
+        return int(cap_value & 0xFFFFFFFF)
+
+    def enumerate_capabilities(self) -> List[str]:
+        """Returns the list of capability strings.
+
+        Returns:
+            List[str]: The list of capability strings.
+        """
+        capabilities_value = self.get_capabilities()
+        return self.capabilities_to_string(capabilities_value)
+
+    def has_capability(self, capability: str) -> bool:
```

Comment:
A lot of this functionality looks like it might already be handled by the [Enumeration](https://github.com/volatilityfoundation/volatility3/blob/8e56cb39d13c22575ba5133ffbc6b83344bed0ba/volatility3/framework/objects/__init__.py#L553) object type, which then stores the constants in the (or an) ISF JSON definition and comes with all this existing functionality: lookup/description/choices/etc.

It would require making a separate JSON file to load the type up and then casting the capability object to that type, but you could also just override the `__new__` and `__init__` methods to provide the data that would come from the ISF to grab it from the constants file instead?

It's not a direct 1-1 mapping from the two, but at the very least it might be good to model the method names on this off of those used for standard volatility enumerations, just for consistency of use.  It looks like this is an array of a capability type, in which case I guess I'd expect each of the array items to be a capability enumeration and then some membership methods in this main class?

Please take a look at the Enumeration class and let me know if you think there's any similarities that could be used, or method names that might help people intuit them when working with the structure.  If not, then no problem, but it feels like there's some commonality there somewhere, so I just want to raise it and investigate...  5:)

Reply:
Ok, I had this ready waiting you reply :) ... Have a look now. It looks much cleaner. I used [dataclasses](https://docs.python.org/3.7/library/dataclasses.html) introduced in Python 3.7 (all good with vol3 minimum requirements) which are really cool, not like namedtuples.